File: runtimes/python/adapter.py
#!/usr/bin/env python3
"""
Unix socket server for Python runtime adapter.
Handles provider discovery and chat requests.
"""

import os
import sys
import socket
import json
import signal
import subprocess
from pathlib import Path
from posthog import Posthog

# Add providers to path
sys.path.insert(0, str(Path(__file__).parent))

from providers import discover_providers

SOCKET_PATH = "/tmp/llm-analytics-python.sock"
PYTHON_RUNTIME_ROOT = Path(__file__).parent
PYTHON_TOOLS_ROOT = PYTHON_RUNTIME_ROOT / "tools"
TOOL_TIMEOUT_SECONDS = 60


class RuntimeAdapter:
    def __init__(self):
        # Environment should be inherited from parent process (Node)
        # Initialize PostHog client
        api_key = os.getenv("POSTHOG_API_KEY")
        if not api_key:
            raise ValueError("POSTHOG_API_KEY not found in environment")

        self.posthog_client = Posthog(
            project_api_key=api_key,
            host=os.getenv("POSTHOG_HOST", "https://app.posthog.com"),
        )

        # Discover providers
        self.providers = {}
        for provider in discover_providers(self.posthog_client):
            name = provider.get_name()
            self.providers[name.lower()] = provider

        print(f"Loaded providers: {list(self.providers.keys())}", file=sys.stderr)
        self.tools = [
            {
                "id": "python_tool_call_selected_provider",
                "name": "Tool Call Smoke Test (Selected Provider)",
                "description": "Runs a weather tool-call test through the active provider.",
            },
            {
                "id": "python_message_selected_provider",
                "name": "Message Smoke Test (Selected Provider)",
                "description": "Runs a simple message test through the active provider.",
            },
            {
                "id": "python_trace_generator",
                "name": "Trace Generator",
                "description": "Runs trace generator in non-interactive mode.",
                "command": sys.executable,
                "args": ["trace_generator.py", "--quick"],
                "cwd": str(PYTHON_TOOLS_ROOT / "trace-generator"),
            },
            {
                "id": "python_screenshot_demo",
                "name": "Screenshot Demo",
                "description": "Runs screenshot demo in non-interactive mode.",
                "command": sys.executable,
                "args": ["screenshot_demo.py", "--tools"],
                "cwd": str(PYTHON_TOOLS_ROOT / "screenshot-demo"),
            },
        ]

    def handle_message(self, message):
        """Handle incoming JSON message"""
        try:
            data = json.loads(message)
            action = data.get("action")

            if action == "get_providers":
                return self.get_providers()
            elif action == "get_provider_options":
                return self.get_provider_options(data.get("provider"))
            elif action == "set_provider_option":
                return self.set_provider_option(
                    data.get("provider"), data.get("option_id"), data.get("value")
                )
            elif action == "chat":
                return self.chat(data.get("provider"), data.get("messages", []))
            elif action == "run_mode_test":
                return self.run_mode_test(data.get("provider"), data.get("mode"))
            elif action == "list_tools":
                return self.list_tools()
            elif action == "run_tool":
                return self.run_tool(data.get("tool_id"), data.get("provider"))
            else:
                return {"error": f"Unknown action: {action}"}

        except Exception as e:
            return {"error": str(e)}

    def get_providers(self):
        """Return list of available providers"""
        providers = [
            {
                "id": name,
                "name": provider.get_name(),
                "options": provider.get_options(),
                "input_modes": provider.get_input_modes(),
            }
            for name, provider in self.providers.items()
        ]
        return {"providers": providers}

    def get_provider_options(self, provider_name):
        """Get options for a specific provider"""
        if not provider_name:
            return {"error": "Provider name required"}

        provider = self.providers.get(provider_name.lower())
        if not provider:
            return {"error": f"Provider not found: {provider_name}"}

        return {"options": provider.get_options()}

    def set_provider_option(self, provider_name, option_id, value):
        """Set an option value for a provider"""
        if not provider_name:
            return {"error": "Provider name required"}

        provider = self.providers.get(provider_name.lower())
        if not provider:
            return {"error": f"Provider not found: {provider_name}"}

        if not option_id:
            return {"error": "Option ID required"}

        try:
            provider.set_option(option_id, value)
            return {"success": True}
        except Exception as e:
            return {"error": f"Failed to set option: {str(e)}"}

    def chat(self, provider_name, messages):
        """Forward chat request to provider"""
        if not provider_name:
            return {"error": "Provider name required"}

        provider = self.providers.get(provider_name.lower())
        if not provider:
            return {"error": f"Provider not found: {provider_name}"}

        try:
            message = provider.chat(messages)
            return {"message": message}
        except Exception as e:
            return {"error": f"Chat failed: {str(e)}"}

    def run_mode_test(self, provider_name, mode):
        if not provider_name:
            return {"error": "Provider name required"}

        provider = self.providers.get(provider_name.lower())
        if not provider:
            return {"error": f"Provider not found: {provider_name}"}

        try:
            if mode == "tool_call_test":
                messages = [{
                    "role": "user",
                    "content": [{"type": "text", "text": "What is the weather in Montreal, Canada?"}]
                }]
                return {"message": provider.chat(messages)}

            if mode == "message_test":
                messages = [{
                    "role": "user",
                    "content": [{"type": "text", "text": "Hi, how are you today?"}]
                }]
                return {"message": provider.chat(messages)}

            if mode == "image_test":
                return {"message": {
                    "role": "assistant",
                    "content": [{"type": "text", "text": "Python runtime image test is not implemented yet."}]
                }}

            if mode == "embeddings_test":
                return {"message": {
                    "role": "assistant",
                    "content": [{"type": "text", "text": "Python runtime embeddings test is not implemented yet."}]
                }}

            if mode == "structured_output_test":
                messages = [{
                    "role": "user",
                    "content": [{"type": "text", "text": "Create a profile for a 25-year-old software developer who loves hiking and photography."}]
                }]
                return {"message": provider.chat(messages)}

            if mode == "transcription_test":
                return {"message": {
                    "role": "assistant",
                    "content": [{"type": "text", "text": "Python runtime transcription test is not implemented yet."}]
                }}

            if mode == "image_generation_test":
                return {"message": {
                    "role": "assistant",
                    "content": [{"type": "text", "text": "Python runtime image generation test is not implemented yet."}]
                }}

            return {"message": {
                "role": "assistant",
                "content": [{"type": "text", "text": f"Unknown mode: {mode}"}]
            }}
        except Exception as e:
            return {"error": f"Mode test failed: {str(e)}"}

    def list_tools(self):
        return {
            "tools": [
                {
                    "id": tool["id"],
                    "name": tool["name"],
                    "description": tool.get("description"),
                }
                for tool in self.tools
            ]
        }

    def run_tool(self, tool_id, provider_name=None):
        if not tool_id:
            return {"error": "tool_id is required"}

        tool = next((candidate for candidate in self.tools if candidate["id"] == tool_id), None)
        if tool is None:
            return {"error": f"Unknown tool: {tool_id}"}

        if tool_id in {"python_tool_call_selected_provider", "python_message_selected_provider"}:
            if not provider_name:
                return {"error": "provider is required for this runtime tool"}

            mode = "tool_call_test" if tool_id == "python_tool_call_selected_provider" else "message_test"
            mode_result = self.run_mode_test(provider_name, mode)
            if "error" in mode_result:
                return mode_result

            message = mode_result.get("message", {})
            text_blocks = [
                block.get("text", "")
                for block in message.get("content", [])
                if block.get("type") == "text"
            ]
            output = "\n".join(text_blocks).strip() or "No output."
            title = "Tool Call Smoke Test" if mode == "tool_call_test" else "Message Smoke Test"

            return {
                "message": {
                    "role": "assistant",
                    "content": [
                        {
                            "type": "text",
                            "text": f"Tool: {title}\nProvider: {provider_name}\n\n{output}",
                        }
                    ],
                }
            }

        try:
            result = subprocess.run(
                [tool["command"], *tool["args"]],
                cwd=tool["cwd"],
                capture_output=True,
                text=True,
                env=os.environ.copy(),
                timeout=TOOL_TIMEOUT_SECONDS,
                check=False,
            )
            timed_out = False
            exit_code = result.returncode
            stdout = result.stdout.strip()
            stderr = result.stderr.strip()
        except subprocess.TimeoutExpired as e:
            timed_out = True
            exit_code = "timeout"
            stdout = (e.stdout or "").strip()
            stderr = (e.stderr or "").strip()
        except Exception as e:
            return {"error": f"Tool execution failed: {str(e)}"}

        output = "\n".join(
            [
                f"Tool: {tool['name']}",
                f"Exit code: {exit_code}{' (timed out)' if timed_out else ''}",
                "",
                f"STDOUT:\n{stdout}" if stdout else "STDOUT: (empty)",
                "",
                f"STDERR:\n{stderr}" if stderr else "STDERR: (empty)",
            ]
        )
        if timed_out:
            output += f"\n\nExecution timed out after {TOOL_TIMEOUT_SECONDS}s."

        return {
            "message": {
                "role": "assistant",
                "content": [{"type": "text", "text": output}],
            }
        }

    def run(self):
        """Start Unix socket server"""
        # Remove old socket if it exists
        if os.path.exists(SOCKET_PATH):
            os.remove(SOCKET_PATH)

        # Create Unix socket
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.bind(SOCKET_PATH)
        sock.listen(1)

        print(f"Listening on {SOCKET_PATH}", file=sys.stderr)

        # Handle shutdown gracefully
        def shutdown(signum, frame):
            print("Shutting down...", file=sys.stderr)
            sock.close()
            if os.path.exists(SOCKET_PATH):
                os.remove(SOCKET_PATH)
            sys.exit(0)

        signal.signal(signal.SIGTERM, shutdown)
        signal.signal(signal.SIGINT, shutdown)

        try:
            while True:
                conn, _ = sock.accept()
                try:
                    # Read message (up to 1MB)
                    data = conn.recv(1024 * 1024).decode("utf-8")
                    if not data:
                        continue

                    # Handle message
                    response = self.handle_message(data)

                    # Send response
                    conn.sendall(json.dumps(response).encode("utf-8"))
                finally:
                    conn.close()

        except Exception as e:
            print(f"Error: {e}", file=sys.stderr)
        finally:
            sock.close()
            if os.path.exists(SOCKET_PATH):
                os.remove(SOCKET_PATH)


if __name__ == "__main__":
    adapter = RuntimeAdapter()
    adapter.run()

File: runtimes/node/adapter.mjs
#!/usr/bin/env node

import fs from "fs";
import os from "os";
import path from "path";
import net from "net";
import { randomUUID } from "crypto";
import { createRequire } from "module";
import { fileURLToPath, pathToFileURL } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const CLI_ROOT = path.resolve(__dirname, "../..");
const LEGACY_NODE_ROOT = path.join(CLI_ROOT, "node");
const PROVIDERS_ROOT = path.join(LEGACY_NODE_ROOT, "dist", "providers");
const SOCKET_PATH = path.join(CLI_ROOT, "runtimes", "node", "adapter.sock");

const requireFromLegacyNode = createRequire(
  path.join(LEGACY_NODE_ROOT, "package.json"),
);
const { PostHog } = requireFromLegacyNode("posthog-node");

const enableSessionId = ["true", "1", "yes"].includes(
  String(process.env.ENABLE_AI_SESSION_ID || "True").toLowerCase(),
);
const aiSessionId = enableSessionId ? randomUUID() : null;

const posthog = new PostHog(process.env.POSTHOG_API_KEY, {
  host: process.env.POSTHOG_HOST || "https://app.posthog.com",
  flushAt: 1,
  flushInterval: 0,
});

const providerDefinitions = [
  {
    id: "anthropic",
    name: "Anthropic",
    moduleFile: "anthropic.js",
    className: "AnthropicProvider",
    streamingModuleFile: "anthropic-streaming.js",
    streamingClassName: "AnthropicStreamingProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
      {
        id: "thinking",
        name: "Extended Thinking",
        shortcutKey: "t",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "gemini",
    name: "Google Gemini",
    moduleFile: "gemini.js",
    className: "GeminiProvider",
    inputModes: ["text", "image", "file"],
    streamingModuleFile: "gemini-streaming.js",
    streamingClassName: "GeminiStreamingProvider",
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "openai_responses",
    name: "OpenAI Responses",
    moduleFile: "openai.js",
    className: "OpenAIProvider",
    inputModes: ["text", "image", "file"],
    streamingModuleFile: "openai-streaming.js",
    streamingClassName: "OpenAIStreamingProvider",
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "openai_chat",
    name: "OpenAI Chat Completions",
    moduleFile: "openai-chat.js",
    className: "OpenAIChatProvider",
    inputModes: ["text", "image", "file"],
    streamingModuleFile: "openai-chat-streaming.js",
    streamingClassName: "OpenAIChatStreamingProvider",
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: true,
      },
    ],
  },
  {
    id: "openai_transcription",
    name: "OpenAI Transcriptions",
    moduleFile: "openai-transcription.js",
    className: "OpenAITranscriptionProvider",
    inputModes: ["audio", "file"],
  },
  {
    id: "openai_image",
    name: "OpenAI Image Generation",
    moduleFile: "openai-image.js",
    className: "OpenAIImageProvider",
    inputModes: ["text"],
  },
  {
    id: "gemini_image",
    name: "Gemini Image Generation",
    moduleFile: "gemini-image.js",
    className: "GeminiImageProvider",
    inputModes: ["text"],
  },
  {
    id: "langchain_openai",
    name: "LangChain (OpenAI)",
    moduleFile: "langchain.js",
    className: "LangChainProvider",
    inputModes: ["text"],
  },
  {
    id: "mastra_openai",
    name: "Mastra (OpenAI)",
    moduleFile: "mastra.js",
    className: "MastraProvider",
    inputModes: ["text"],
  },
  {
    id: "vercel_ai_openai",
    name: "Vercel AI SDK (OpenAI)",
    moduleFile: "vercel-ai.js",
    className: "VercelAIProvider",
    streamingModuleFile: "vercel-ai-streaming.js",
    streamingClassName: "VercelAIStreamingProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "vercel_ai_google",
    name: "Vercel AI SDK (Google)",
    moduleFile: "vercel-ai-google.js",
    className: "VercelAIGoogleProvider",
    streamingModuleFile: "vercel-ai-google-streaming.js",
    streamingClassName: "VercelAIGoogleStreamingProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "vercel_ai_anthropic",
    name: "Vercel AI SDK (Anthropic)",
    moduleFile: "vercel-ai-anthropic.js",
    className: "VercelAIAnthropicProvider",
    streamingModuleFile: "vercel-ai-anthropic-streaming.js",
    streamingClassName: "VercelAIAnthropicStreamingProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "vercel_ai_gateway_anthropic",
    name: "Vercel AI Gateway (Anthropic)",
    moduleFile: "vercel-ai-gateway-anthropic.js",
    className: "VercelAIGatewayAnthropicProvider",
    streamingModuleFile: "vercel-ai-gateway-anthropic-streaming.js",
    streamingClassName: "VercelAIGatewayAnthropicStreamingProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: false,
      },
    ],
  },
  {
    id: "vercel_generate_object",
    name: "Vercel AI SDK - generateObject (OpenAI)",
    moduleFile: "vercel-generate-object.js",
    className: "VercelGenerateObjectProvider",
    inputModes: ["text", "image", "file"],
  },
  {
    id: "vercel_stream_object",
    name: "Vercel AI SDK - streamObject (OpenAI)",
    moduleFile: "vercel-stream-object.js",
    className: "VercelStreamObjectProvider",
    streamingModuleFile: "vercel-stream-object.js",
    streamingClassName: "VercelStreamObjectProvider",
    inputModes: ["text", "image", "file"],
    options: [
      {
        id: "streaming",
        name: "Streaming",
        shortcutKey: "s",
        type: "boolean",
        default: true,
      },
    ],
  },
];

const providerOptions = new Map();
const providerInstances = new Map();

for (const def of providerDefinitions) {
  const defaults = {};
  for (const option of def.options || []) {
    defaults[option.id] = option.default;
  }
  providerOptions.set(def.id, defaults);
}

function getProviderDefinition(providerId) {
  const definition = providerDefinitions.find((def) => def.id === providerId);
  if (!definition) {
    throw new Error(`Unknown provider: ${providerId}`);
  }
  return definition;
}

async function instantiateProvider(providerId, forceStreaming = false) {
  const definition = getProviderDefinition(providerId);
  const options = providerOptions.get(providerId) || {};
  const useStreaming =
    (forceStreaming || Boolean(options.streaming)) &&
    definition.streamingModuleFile &&
    definition.streamingClassName;

  const moduleFile = useStreaming
    ? definition.streamingModuleFile
    : definition.moduleFile;
  const className = useStreaming
    ? definition.streamingClassName
    : definition.className;

  const modulePath = path.join(PROVIDERS_ROOT, moduleFile);
  const providerModule = await import(pathToFileURL(modulePath).href);
  const ProviderClass = providerModule[className];

  if (!ProviderClass) {
    throw new Error(
      `Provider class "${className}" not found in ${moduleFile}`,
    );
  }

  if (providerId.startsWith("anthropic")) {
    return new ProviderClass(posthog, Boolean(options.thinking), undefined, aiSessionId);
  }

  return new ProviderClass(posthog, aiSessionId);
}

async function getProvider(providerId) {
  const cached = providerInstances.get(providerId);
  if (cached) {
    return cached;
  }

  const instance = await instantiateProvider(providerId);
  providerInstances.set(providerId, instance);
  return instance;
}

async function getStreamingProvider(providerId) {
  const definition = getProviderDefinition(providerId);
  if (!definition.streamingModuleFile || !definition.streamingClassName) {
    return getProvider(providerId);
  }

  return instantiateProvider(providerId, true);
}

function getLatestUserInput(messages) {
  const userMessage = [...messages].reverse().find((msg) => msg.role === "user");
  if (!userMessage) {
    return { text: "Hello!", imageBase64: undefined };
  }

  const textParts = [];
  let imageBase64;

  for (const block of userMessage.content || []) {
    if (block.type === "text" && block.text) {
      textParts.push(block.text);
      continue;
    }

    if (imageBase64) {
      continue;
    }

    if (
      block.type === "image" ||
      (block.type === "file" &&
        typeof block.mimeType === "string" &&
        block.mimeType.startsWith("image/"))
    ) {
      try {
        const bytes = fs.readFileSync(block.path);
        imageBase64 = bytes.toString("base64");
      } catch (err) {
        console.error(`Failed to read image file ${block.path}:`, err);
      }
    }
  }

  const text = textParts.join("\n").trim() || "Please describe this image.";
  return { text, imageBase64 };
}

function createTextMessage(text) {
  return {
    role: "assistant",
    content: [{ type: "text", text }],
  };
}

async function runModeTest(providerId, mode) {
  const provider = await getProvider(providerId);

  if (typeof provider.resetConversation === "function") {
    provider.resetConversation();
  }

  if (mode === "tool_call_test") {
    const response = await provider.chat("What is the weather in Montreal, Canada?");
    return createTextMessage(response);
  }

  if (mode === "message_test") {
    const response = await provider.chat("Hi, how are you today?");
    return createTextMessage(response);
  }

  if (mode === "image_test") {
    const base64Image =
      "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z8DwHwAFBQIAX8jx0gAAAABJRU5ErkJggg==";
    const response = await provider.chat(
      "What do you see in this image? Please describe it.",
      base64Image,
    );
    return createTextMessage(response);
  }

  if (mode === "embeddings_test") {
    if (typeof provider.embed !== "function") {
      return createTextMessage(`Provider ${providerId} does not support embeddings.`);
    }
    const embedding = await provider.embed("The quick brown fox jumps over the lazy dog.");
    const sample = Array.isArray(embedding) ? embedding.slice(0, 5).join(", ") : "";
    return createTextMessage(
      `Embedding generated (${embedding?.length || 0} dimensions). Sample: [${sample}]`,
    );
  }

  if (mode === "structured_output_test") {
    const response = await provider.chat(
      "Create a profile for a 25-year-old software developer who loves hiking and photography.",
    );
    return createTextMessage(response);
  }

  if (mode === "transcription_test") {
    if (typeof provider.transcribe !== "function") {
      return createTextMessage(`Provider ${providerId} does not support transcription.`);
    }
    const audioPath = path.join(CLI_ROOT, "test-audio.mp3");
    if (!fs.existsSync(audioPath)) {
      return createTextMessage(`Test audio file not found: ${audioPath}`);
    }
    const text = await provider.transcribe(audioPath);
    return createTextMessage(text || "No transcription generated.");
  }

  if (mode === "image_generation_test") {
    if (typeof provider.generateImage !== "function") {
      return createTextMessage(`Provider ${providerId} does not support image generation.`);
    }
    const imageResult = await provider.generateImage(
      "A serene mountain landscape at sunset with a lake reflection",
    );
    return createTextMessage(imageResult || "No image result returned.");
  }

  return createTextMessage(`Unknown mode: ${mode}`);
}

async function handleMessage(rawData) {
  const data = JSON.parse(rawData.toString("utf-8"));
  const action = data.action;

  if (action === "get_providers") {
    return {
      providers: providerDefinitions.map((provider) => ({
        id: provider.id,
        name: provider.name,
        options: provider.options || [],
        input_modes: provider.inputModes,
      })),
    };
  }

  if (action === "set_provider_option") {
    const { provider, option_id, value } = data;
    const definition = getProviderDefinition(provider);
    const optionDef = (definition.options || []).find((opt) => opt.id === option_id);

    if (!optionDef) {
      throw new Error(`Unknown option "${option_id}" for provider "${provider}"`);
    }

    const options = providerOptions.get(provider) || {};
    options[option_id] = value;
    providerOptions.set(provider, options);
    providerInstances.delete(provider);

    return { success: true };
  }

  if (action === "chat") {
    const providerId = data.provider;
    const messages = data.messages || [];

    const provider = await getProvider(providerId);
    const { text, imageBase64 } = getLatestUserInput(messages);

    let responseText = "";

    if (typeof provider.chatStream === "function" && !imageBase64) {
      for await (const chunk of provider.chatStream(text)) {
        responseText += chunk;
      }
    } else {
      responseText = await provider.chat(text, imageBase64);
    }

    return {
      message: {
        role: "assistant",
        content: [{ type: "text", text: responseText }],
      },
    };
  }

  if (action === "run_mode_test") {
    const providerId = data.provider;
    const mode = data.mode;
    const message = await runModeTest(providerId, mode);
    return { message };
  }

  throw new Error(`Unknown action: ${action}`);
}

async function handleStreamingRequest(conn, data) {
  const providerId = data.provider;
  const messages = data.messages || [];

  const provider = await getStreamingProvider(providerId);
  const { text, imageBase64 } = getLatestUserInput(messages);

  let responseText = "";

  if (typeof provider.chatStream === "function" && !imageBase64) {
    for await (const chunk of provider.chatStream(text)) {
      responseText += chunk;
      conn.write(JSON.stringify({ type: "chunk", chunk }) + "\n");
    }
  } else {
    responseText = await provider.chat(text, imageBase64);
    conn.write(JSON.stringify({ type: "chunk", chunk: responseText }) + "\n");
  }

  conn.write(
    JSON.stringify({
      type: "done",
      message: {
        role: "assistant",
        content: [{ type: "text", text: responseText }],
      },
    }) + "\n",
  );
}

try {
  fs.mkdirSync(path.dirname(SOCKET_PATH), { recursive: true });
  if (fs.existsSync(SOCKET_PATH)) {
    fs.unlinkSync(SOCKET_PATH);
  }

  const server = net.createServer({ allowHalfOpen: true }, (conn) => {
    let data = "";

    conn.on("data", (chunk) => {
      data += chunk.toString();
    });

    conn.on("end", async () => {
      if (!data.trim()) {
        conn.write(JSON.stringify({ error: "Empty request payload" }));
        conn.end();
        return;
      }

      try {
        const parsed = JSON.parse(data);
        if (parsed.action === "chat_stream") {
          await handleStreamingRequest(conn, parsed);
        } else {
          const response = await handleMessage(data);
          conn.write(JSON.stringify(response));
        }
      } catch (err) {
        const message = err instanceof Error ? err.message : String(err);
        if (data.includes('"action":"chat_stream"')) {
          conn.write(JSON.stringify({ type: "error", error: message }) + "\n");
        } else {
          conn.write(JSON.stringify({ error: message }));
        }
      } finally {
        conn.end();
      }
    });

    conn.on("error", (err) => {
      console.error("Connection error:", err);
    });
  });

  server.listen(SOCKET_PATH, () => {
    console.error(`Node runtime listening on ${SOCKET_PATH}`);
  });

  const shutdown = () => {
    try {
      server.close();
    } finally {
      try {
        posthog.shutdown();
      } catch {
        // no-op
      }
      if (fs.existsSync(SOCKET_PATH)) {
        fs.unlinkSync(SOCKET_PATH);
      }
      process.exit(0);
    }
  };

  process.on("SIGINT", shutdown);
  process.on("SIGTERM", shutdown);
} catch (err) {
  console.error(err);
  process.exit(1);
}
